airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py

```diff
@@ -0,0 +1,394 @@
+        }
+
+        resp = requests.post(self.BASE_URL + "/oauth/v1/token", data=payload)
+        if resp.status_code == 403:
+            raise HubspotInvalidAuth(resp.content, response=resp)
+
+        resp.raise_for_status()
+        auth = resp.json()
+        self._credentials["access_token"] = auth["access_token"]
+        self._credentials["refresh_token"] = auth["refresh_token"]
+        self._credentials["token_expires"] = datetime.utcnow() + timedelta(seconds=auth["expires_in"] - 600)
+        logger.info("Token refreshed. Expires at %s", self._credentials["token_expires"])
+
+    @property
+    def api_key(self) -> Optional[str]:
+        """Get API Key if set"""
+        return self._credentials.get("api_key")
+
+    @property
+    def access_token(self) -> Optional[str]:
+        """Get Access Token if set, refreshes token if needed"""
+        if not self._credentials.get("access_token"):
+            return None
+
+        if self._credentials["token_expires"] is None or self._credentials["token_expires"] < datetime.utcnow():
+            self._acquire_access_token_from_refresh_token()
+        return self._credentials.get("access_token")
+
+    def _add_auth(self, params: Mapping[str, Any] = None) -> Mapping[str, Any]:
+        """Add auth info to request params/header"""
+        params = params or {}
+
+        if self.api_key:
+            params["hapikey"] = self.api_key
+        else:
+            self._session.headers["Authorization"] = f"Bearer {self.access_token}"
+
+        return params
+
+    @staticmethod
+    def _parse_and_handle_errors(response) -> Mapping[str, Any]:
+        """Handle response"""
+        if response.status_code == 403:
+            raise HubspotSourceUnavailable(response.content)
+        elif response.status_code == 429:
+            retry_after = response.headers.get("Retry-After")
+            raise HubspotRateLimited(
+                f"429 Rate Limit Exceeded: API rate-limit has been reached until {retry_after} seconds."
+                " See https://developers.hubspot.com/docs/api/usage-details",
+                response=response,
+            )
+        else:
+            response.raise_for_status()
+
+        return response.json()
+
+    @retry_connection_handler(max_tries=5, factor=5)
+    @retry_after_handler(max_tries=3)
+    def get(self, url: str, params=None) -> Union[Mapping[str, Any], List[Mapping[str, Any]]]:
+        response = self._session.get(self.BASE_URL + url, params=self._add_auth(params))
+        return self._parse_and_handle_errors(response)
+
+    def post(self, url: str, data: Mapping[str, Any], params=None) -> Union[Mapping[str, Any], List[Mapping[str, Any]]]:
+        response = self._session.post(self.BASE_URL + url, params=self._add_auth(params), json=data)
+        return self._parse_and_handle_errors(response)
+
+
+class StreamAPI(ABC):
+    entity = None
+
+    more_key = None
+    data_path = "results"
```

Comment:
`data_path` is a bit misleading as a name since it's a single key, not a path in the JSON. Can we use something like `data_field`?